// @angular-cool/storage/esm5/lib/cool-storage-base.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Subject } from 'rxjs';
/**
 * @abstract
 */
var CoolStorageBase = /** @class */ (function () {
    /**
     * @param {?} storageObject
     * @param {?} storageObjectName
     */
    function CoolStorageBase(storageObject, storageObjectName) {
        if (!storageObject) {
            throw new Error("Current browser does not support " + storageObjectName);
        }
        this._storageObject = storageObject;
        this._itemSetSubject = new Subject();
        this._itemRemovedSubject = new Subject();
    }
    Object.defineProperty(CoolStorageBase.prototype, "itemSetObservable", {
        /**
         * @return {?}
         */
        get: function () {
            return this._itemSetSubject.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CoolStorageBase.prototype, "itemRemovedObservable", {
        /**
         * @return {?}
         */
        get: function () {
            return this._itemRemovedSubject.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} key
     * @return {?}
     */
    CoolStorageBase.prototype.getItem = function (key) {
        return this._storageObject.getItem(key) || null;
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    CoolStorageBase.prototype.setItem = function (key, value) {
        this._itemSetSubject.next({
            key: key,
            value: value
        });
        this._setItemInStorage(key, value);
    };
    /**
     * @param {?} key
     * @return {?}
     */
    CoolStorageBase.prototype.removeItem = function (key) {
        var /** @type {?} */ currentValue = this.tryGetObject(key);
        if (!currentValue) {
            currentValue = this.getItem(key);
        }
        this._itemRemovedSubject.next({
            key: key,
            value: currentValue
        });
        this._storageObject.removeItem(key);
    };
    /**
     * @param {?} index
     * @return {?}
     */
    CoolStorageBase.prototype.key = function (index) {
        return this._storageObject.key(index);
    };
    /**
     * @return {?}
     */
    CoolStorageBase.prototype.clear = function () {
        this._storageObject.clear();
    };
    Object.defineProperty(CoolStorageBase.prototype, "length", {
        /**
         * @return {?}
         */
        get: function () {
            return this._storageObject.length;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @template T
     * @param {?} key
     * @return {?}
     */
    CoolStorageBase.prototype.getObject = function (key) {
        var /** @type {?} */ jsonInStorage = this.getItem(key);
        if (jsonInStorage === null) {
            return null;
        }
        return /** @type {?} */ (JSON.parse(jsonInStorage));
    };
    /**
     * @template T
     * @param {?} key
     * @return {?}
     */
    CoolStorageBase.prototype.tryGetObject = function (key) {
        try {
            return this.getObject(key);
        }
        catch ( /** @type {?} */e) {
            return null;
        }
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    CoolStorageBase.prototype.setObject = function (key, value) {
        this._itemSetSubject.next({
            key: key,
            value: value
        });
        this._setItemInStorage(key, JSON.stringify(value));
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    CoolStorageBase.prototype._setItemInStorage = function (key, value) {
        this._storageObject.setItem(key, value);
    };
    return CoolStorageBase;
}());
export { CoolStorageBase };
function CoolStorageBase_tsickle_Closure_declarations() {
    /** @type {?} */
    CoolStorageBase.prototype._storageObject;
    /** @type {?} */
    CoolStorageBase.prototype._itemSetSubject;
    /** @type {?} */
    CoolStorageBase.prototype._itemRemovedSubject;
}
