// @angular-cool/storage/esm5/lib/cool-server-storage.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var CoolServerStorage = /** @class */ (function () {
    function CoolServerStorage() {
        this._storageObject = {};
    }
    /**
     * @param {?} key
     * @return {?}
     */
    CoolServerStorage.prototype.getItem = function (key) {
        return this._storageObject[key] || null;
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    CoolServerStorage.prototype.setItem = function (key, value) {
        this._storageObject[key] = value;
    };
    /**
     * @param {?} key
     * @return {?}
     */
    CoolServerStorage.prototype.removeItem = function (key) {
        this._storageObject[key] = undefined;
    };
    /**
     * @param {?} index
     * @return {?}
     */
    CoolServerStorage.prototype.key = function (index) {
        return this._storageObject.key(index);
    };
    /**
     * @return {?}
     */
    CoolServerStorage.prototype.clear = function () {
        this._storageObject = {};
    };
    Object.defineProperty(CoolServerStorage.prototype, "length", {
        /**
         * @return {?}
         */
        get: function () {
            return Object.keys(this._storageObject).length;
        },
        enumerable: true,
        configurable: true
    });
    return CoolServerStorage;
}());
export { CoolServerStorage };
function CoolServerStorage_tsickle_Closure_declarations() {
    /** @type {?} */
    CoolServerStorage.prototype._storageObject;
}
