// @angular-cool/http/esm5/lib/cool-http.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { HttpClientModule } from '@angular/common/http';
import { CoolHttp } from './cool-http.service';
var CoolHttpModule = /** @class */ (function () {
    function CoolHttpModule() {
    }
    /**
     * @deprecated
     * @return {?}
     */
    CoolHttpModule.forRoot = function () {
        return {
            ngModule: CoolHttpModule,
            providers: []
        };
    };
    return CoolHttpModule;
}());
export { CoolHttpModule };
CoolHttpModule.decorators = [
    { type: NgModule, args: [{
                exports: [],
                imports: [HttpClientModule],
                providers: [CoolHttp]
            },] },
];
function CoolHttpModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    CoolHttpModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    CoolHttpModule.ctorParameters;
}
