// @angular-cool/storage/esm2015/lib/cool-server-storage.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
export class CoolServerStorage {
    constructor() {
        this._storageObject = {};
    }
    /**
     * @param {?} key
     * @return {?}
     */
    getItem(key) {
        return this._storageObject[key] || null;
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    setItem(key, value) {
        this._storageObject[key] = value;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    removeItem(key) {
        this._storageObject[key] = undefined;
    }
    /**
     * @param {?} index
     * @return {?}
     */
    key(index) {
        return this._storageObject.key(index);
    }
    /**
     * @return {?}
     */
    clear() {
        this._storageObject = {};
    }
    /**
     * @return {?}
     */
    get length() {
        return Object.keys(this._storageObject).length;
    }
}
function CoolServerStorage_tsickle_Closure_declarations() {
    /** @type {?} */
    CoolServerStorage.prototype._storageObject;
}
